docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,76 @@
+scenarios is not implemented.
+
+## Idea
+
+Design is heavily influenced by peculiarities of the upstream and downstream systems
+about which I know very little.
+
+### Downstream
+
+The downstream system is DataSet REST API and its API call [addEvents](https://app.scalyr.com/help/api#addEvents).
+It has the following properties:
+
+* `D1` - only one request for each session could be made
+* `D2` - when new version is deployed there is few minutes outage
+* `D3` - maximum payload size for the API is 6MB before compression
+
+### Upstream
+
+The upstream system is [open telemetry collector](https://opentelemetry.io/docs/collector/),
+where the datasetexporter is one of many exporters that could be used
+(on the right side of the image). It has the following properties:
+
+* `U1` - data are passed in batches
+* `U2` - each batch could be either accepted or rejected, if it's accepted it cannot be returned, if it's rejected, it may be retried
+* `U3` - function for accepting batches could be called in parallel (with different batches)
+
+### Consequences
+
+As a consequence of the upstream and downstream system following has to happen (codes
+between brackets points to the previous points):
+
+* `C1` - data has to be grouped to be divided into multiple sessions to increase throughput (`D1`)
+* `C2` - data from multiple input batches has to be combined (`D1`, `U1`)
+* `C3` - input and output has to be decoupled (`D1`, `C1`, `C2`)
+* `C4` - if the downstream returns error, system cannot return data that has already accepted, it can only start rejecting new data (`U2`, `U3`)
+* `C5` - since the system cannot return data, it has to handle retries by itself (`C4`)
+* `C6` - user has to be able to define the grouping mechanism to achieve maximum throughput (`C1`)
+
+## Implementation
+
+* `I1` - to decouple upstream from downstream we use pub/sub pattern (`C3`)
+* `I2` - there are two levels of pub/sub patterns
+  * `I2.1` - publisher 1 - events from incoming batch are divided based on the grouping into sessions and published into the corresponding topic (`I1`, `C6`)
+  * `I2.2` - subscriber 1 / publisher 2 - reads individual events from the topic and combines them into batches, batches are published into the corresponding topic (`C2`)
+  * `I2.3` - publisher 2 - output batch is published even if it's not full after some time period
+  * `I2.4` - subscriber 2 - reads output batches from the topic and calls DataSet API server (`D1`)
+* `I3` - each session has its own go routine (`I2.1`)
+
+### Error Propagation
+
+Error state is shared among all sessions. It works in the following way:
+
+* when error is returned from the API call then:
+  * system enters error mode
+  * internal retry mechanism is triggered
+* when the system is in error mode, then:
+  * new incoming batches are rejected so that external retry mechanism is triggered
+  * internal retry are still happening
+* error state is cleared when:
+  * some of the internal retries succeeds
```

Comment:
I assume sessions send batches in parallel. So when SessionA triggers and error mode and no new batches are consumed for any session, but SessionB still may send request (previously published in topics) to DataSet. What if also sessionB request fails as well. And than SessionA retry succeeds. What happens?

`some of the internal retries succeeds` does internal mean session scope? Or any success returns system to OK mode. No matter of SessionB still retries (still in error mode)?